xdsl/ir.py

```diff
@@ -294,12 +294,34 @@ def __str__(self) -> str:
 
 DataElement = TypeVar("DataElement")
 
+_D = TypeVar("_D", bound="Data[Any]")
+
 
 @dataclass(frozen=True)
 class Data(Generic[DataElement], Attribute, ABC):
     """An attribute represented by a Python structure."""
     data: DataElement
 
+    @classmethod
+    def new(cls: type[_D], params: Any) -> _D:
+        """
+        Create a new `Data` given its parameter.
+
+        The `params` argument should be of the same type as the `Data` generic
+        argument.
```

Comment:
Can't we enforce this in the type hints?

Reply:
Not as far as I know, since pyright (or Python) disallow using `TypeVar` in a `TypeVar`, so we cannot express the `cls` type.